docs/guide/tutorials/gridfs.txt

```diff
@@ -1,3 +1,206 @@
+
+:manual:`GridFS </core/gridfs/>` is a specification for storing and 
+retrieving files that exceed the :manual:`BSON-document size limit </reference/limits/#limit-bson-document-size>` 
+of 16 megabytes.
+
+Instead of storing a file in a single document, GridFS divides a file into parts, or chunks,
+and stores each of those chunks as a separate document. By default, GridFS limits chunk size
+to 255 kilobytes. GridFS uses two collections to store files: the ``chunks`` collection which
+stores the file chunks, and the ``files`` collection that stores the file metadata.
+
+When you query a GridFS store for a file, the driver or client will reassemble the chunks as 
+needed. GridFS is useful not only for storing files that exceed 16 megabytes but also for
+storing any files which you want to access without having to load the entire file into memory.
+
+The Node Driver supports GridFS with an api that is compatible with
+`Node Streams <https://nodejs.org/dist/latest/docs/api/stream.html>`_ , so you can ``.pipe()`` 
+directly from file streams to MongoDB. In this tutorial, you will see how to use the GridFS
+streaming API to upload
+`a CC-licensed 28 MB recording of the overture from Richard Wagner's opera *Die Meistersinger von Nurnberg* <https://musopen.org/music/213/richard-wagner/die-meistersinger-von-nurnberg-overture/>`_
+to MongoDB using streams.
+
+Uploading a File
+----------------
+
+You can use GridFS to upload a file to MongoDB. This example
+assumes that you have a file named ``meistersinger.mp3`` in the
+root directory of your project. You can use whichever file you want, or you
+can just download a `\ *Die Meistersinger* Overture mp3 <https://musopen.org/music/213/richard-wagner/die-meistersinger-von-nurnberg-overture/>`_.
+
+In order to use the streaming GridFS API, you first need to create
+a ``GridFSBucket``.
+
+.. code-block:: js
+
+   const { MongoClient, GridFSBucket } = require('mongodb');
+
+   const uri = 'mongodb://localhost:27017';
+   const dbName = 'test';
+
+   const client = new MongoClient(uri);
+
+   client.connect().then(async function() {
+     try {
+       const db = client.db(dbName);
+       const bucket = new GridFSBucket(db);
+
+       // Use bucket...
+     } finally {
+       await client.close();
+     }
+   });
+
+The bucket has an ``openUploadStream()`` method that creates an upload stream for a given
+file name. You can pipe a Node.js ``fs`` read stream to the upload stream.
+
+.. code-block:: js
+
+   const fs = require('fs');
+   const { MongoClient, GridFSBucket } = require('mongodb');
+
+   const uri = 'mongodb://localhost:27017';
+   const dbName = 'test';
+
+   const client = new MongoClient(uri);
+
+   client.connect().then(async function() {
+     try {
+       const db = client.db(dbName);
+       const bucket = new GridFSBucket(db);
+
+       // We need to make sure we wait for finish or error,
+       // so we wrap this all in a Promise.
+
+       await new Promise((resolve, reject) => {
```

Comment:
If I was writing this as a command line utility I don't think I'd write it like this, right? It seems like we're trying to "round peg square hole" a stream into an awaitable promise. Remember that people will just straight up copy and paste this code into their applications, so I think a more "streamy" example for this one is apt.

Reply:
honestly, I would probably write it like this. It's not great, but we don't really have a good paradigm for doing streaming with our client.

Here is what the example would look like if we did callbacks:

```js

function finish(runError) {
  client.close(closeError => {
    const err = runError || closeError;
    if (err) {
      console.log(err);
      process.exit(1);
    }
    process.exit(0);
  });
}

client.connect(err => {
  if (err) {
    return finish(err);
  }

  const db = client.db(dbName);
  const bucket = new GridFSBucket(db);

  fs.createReadStream('./meistersinger.mp3')
    .pipe(bucket.openUploadStream('meistersinger.mp3'))
    .on('error', finish)
    .on('finish', () => finish());
});
```

The code looks neater at the actual example of the gridfs stuff, but handling the `client.close` and the error stuff is more complicated.

Reply:
This looks way better to me!